openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+                      'enrichments. Consider setting the isotopic ' \
+                      'composition manually for enrichments over 5%.'.\
+                      format(enrichment, self._id)
+                warnings.warn(msg)
+
+        # Make sure element name is just that
+        if not element.isalpha():
+            raise ValueError("Element name should be given by the "
+                             "element's symbol, e.g., 'Zr'")
+
+        # Add naturally-occuring isotopes
+        element = openmc.Element(element)
+        for nuclide in element.expand(percent, percent_type, enrichment):
+            self.add_nuclide(*nuclide)
+
+    def add_s_alpha_beta(self, name, fraction=1.0):
+        r"""Add an :math:`S(\alpha,\beta)` table to the material
+
+        Parameters
+        ----------
+        name : str
+            Name of the :math:`S(\alpha,\beta)` table
+        fraction : float
+            The fraction of relevant nuclei that are affected by the
+            :math:`S(\alpha,\beta)` table.  For example, if the material is a
+            block of carbon that is 60% graphite and 40% amorphous then add a
+            graphite :math:`S(\alpha,\beta)` table with fraction=0.6.
+
+        """
+
+        if self._macroscopic is not None:
+            msg = 'Unable to add an S(a,b) table to Material ID="{}" as a ' \
+                  'macroscopic data-set has already been added'.format(self._id)
+            raise ValueError(msg)
+
+        if not isinstance(name, str):
+            msg = 'Unable to add an S(a,b) table to Material ID="{}" with a ' \
+                        'non-string table name "{}"'.format(self._id, name)
+            raise ValueError(msg)
+
+        cv.check_type('S(a,b) fraction', fraction, Real)
+        cv.check_greater_than('S(a,b) fraction', fraction, 0.0, True)
+        cv.check_less_than('S(a,b) fraction', fraction, 1.0, True)
+
+        new_name = openmc.data.get_thermal_name(name)
+        if new_name != name:
+            msg = 'OpenMC S(a,b) tables follow the GND naming convention. ' \
+                  'Table "{}" is being renamed as "{}".'.format(name, new_name)
+            warnings.warn(msg)
+
+        self._sab.append((new_name, fraction))
+
+    def make_isotropic_in_lab(self):
+        self.isotropic = [x[0] for x in self._nuclides]
+
+    def get_nuclides(self):
+        """Returns all nuclides in the material
+
+        Returns
+        -------
+        nuclides : list of str
+            List of nuclide names
+
+        """
+        return [x[0] for x in self._nuclides]
+
+    def get_nuclide_densities(self):
+        """Returns all nuclides in the material and their densities
+
+        Returns
+        -------
+        nuclides : dict
+            Dictionary whose keys are nuclide names and values are 3-tuples of
+            (nuclide, density percent, density percent type)
+
+        """
+
+        nuclides = OrderedDict()
```

Comment:
consider a comment about why the order matters here (oftentimes it's just for stable behavior and/or testing but that's important behavior to note and retain if this code is ever refactored)